sgl-kernel/csrc/cpu/gemm_fp8.cpp

```diff
@@ -0,0 +1,356 @@
+      data_index_init(begin, mb, MB, nb, NB);
+
+      // for brgemm, use float32 for accumulate
+      alignas(64) float Ctmp[BLOCK_M * BLOCK_N];
+      // for brgemm when mat2 is float8_e4m3
+      alignas(64) scalar_t Btmp[BLOCK_N * BLOCK_K];
+
+      for (int64_t i = begin; i < end; ++i) {
+        UNUSED(i);
+        const float* scale_ptr = scales2 + (nb / blocks_n_per_group) * scale_size_K;
+
+        int64_t mb_start = mb * BLOCK_M;
+        int64_t mb_size = std::min(M - mb_start, BLOCK_M);
+        int64_t nb_start = nb * BLOCK_N;
+        int64_t nb_size = std::min(N - nb_start, BLOCK_N);
+
+        tinygemm_kernel<scalar_t, packed_t, has_bias>(
+            /*   A                  */ mat1 + mb_start * mat1_strideM,
+            /*   B                  */ mat2 + nb_start * K, // nb * BLOCK_N * K
+            /*   C                  */ out + mb_start * out_strideM + nb_start,
+            /*   Btmp               */ Btmp,
+            /*   Ctmp               */ Ctmp,
+            /*   scale              */ scale_ptr,
+            /*   bias               */ bias + nb_start,
+            /*   M                  */ mb_size,
+            /*   N                  */ nb_size,
+            /*   K                  */ K,
+            /*   lda                */ mat1_strideM,
+            /*   ldb                */ nb_size,
+            /*   ldc                */ out_strideM,
+            /*   brg                */ use_brgemm,
+            /*   blocks_k_per_group */ blocks_k_per_group,
+            /*   block_size_K       */ block_size_K);
+
+        // move to the next index
+        data_index_step(mb, MB, nb, NB);
+      }
+
+      if (use_brgemm) {
+        at::native::cpublas::brgemm_release();
+      }
+    });
+  });
+
+}
+
+} // anonymous namespace
+
+at::Tensor fp8_scaled_mm_cpu(at::Tensor& mat1, at::Tensor& mat2, at::Tensor& scales2,
+    std::vector<int64_t> block_size, std::optional<at::Tensor>& bias,
+    at::ScalarType out_dtype, bool is_vnni) {
+  RECORD_FUNCTION("sgl-kernel::fp8_scaled_mm_cpu", std::vector<c10::IValue>({mat1, mat2, scales2, block_size, bias}));
+
+  auto packed_w = is_vnni ? mat2 : convert_weight_packed(mat2);
+
+  CHECK_LAST_DIM_CONTIGUOUS_INPUT(mat1);
+  CHECK_INPUT(mat2);
+  CHECK_INPUT(scales2);
+  TORCH_CHECK(scales2.scalar_type() == at::kFloat,
+      "fp8_scaled_mm_cpu: expect scales2 to be float32.");
+
+  int64_t M = mat1.size(0);
+  int64_t N = mat2.size(0);
+  int64_t K = mat2.size(1);
+
+  CHECK_EQ(mat1.size(1), K);
+  CHECK_DIM(2, mat1);
+  CHECK_DIM(2, mat2);
+
+  TORCH_CHECK(block_size.size() == 2,
+      "fp8_scaled_mm_cpu: expect block_size.size() to be 2.");
+
+  int64_t block_size_N = block_size[0];
+  int64_t block_size_K = block_size[1];
+
+  constexpr int64_t BLOCK_N = block_size_n();
+  TORCH_CHECK(block_size_N >= BLOCK_N, "fp8_scaled_mm_cpu: expect block_size_N >= BLOCK_N");
+  TORCH_CHECK(block_size_K >= BLOCK_K, "fp8_scaled_mm_cpu: expect block_size_K >= BLOCK_K");
```

Comment:
```suggestion
  TORCH_CHECK(block_size_N % BLOCK_N == 0, "fp8_scaled_mm_cpu: expect block_size_N to be multiples of BLOCK_N");
  TORCH_CHECK(block_size_K == BLOCK_K, "fp8_scaled_mm_cpu: expect block_size_K equals to BLOCK_K");
```

this PR won't support random block_size_K here. for example if BLOCK_K is 96 and block_size_K is 128, then the loop
```
int k = 0; k < K; k += BLOCK_K
```
will end up with 2 different scales in each 128.

but i suppose we don't have to be that complexed, just make sure that BLOCK_K is the same value as block_size_K, that will be all right.

Reply:
Updated in https://github.com/mingfeima/sglang/pull/47/commits/9415c82420d36e680976ef72a7389ab31b1fc363